airflow/executors/kubernetes_executor_utils.py

```diff
@@ -0,0 +1,466 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+import json
+import multiprocessing
+import time
+from queue import Empty, Queue
+from typing import TYPE_CHECKING, Any
+
+from kubernetes import client, watch
+from kubernetes.client import Configuration, models as k8s
+from kubernetes.client.rest import ApiException
+from urllib3.exceptions import ReadTimeoutError
+
+from airflow.exceptions import AirflowException
+from airflow.kubernetes.kube_client import get_kube_client
+from airflow.kubernetes.kubernetes_helper_functions import annotations_to_key, create_pod_id
+from airflow.kubernetes.pod_generator import PodGenerator
+from airflow.utils.log.logging_mixin import LoggingMixin
+from airflow.utils.state import State
+
+if TYPE_CHECKING:
+    from airflow.executors.kubernetes_executor_types import (
+        KubernetesJobType,
+        KubernetesResultsType,
+        KubernetesWatchType,
+    )
+
+
+from airflow.executors.kubernetes_executor_types import ALL_NAMESPACES, POD_EXECUTOR_DONE_KEY
+
+
+class ResourceVersion:
+    """Singleton for tracking resourceVersion from Kubernetes."""
+
+    _instance: ResourceVersion | None = None
+    resource_version: dict[str, str] = {}
+
+    def __new__(cls):
+        if cls._instance is None:
+            cls._instance = super().__new__(cls)
+        return cls._instance
+
+
+class KubernetesJobWatcher(multiprocessing.Process, LoggingMixin):
```

Comment:
What do folks think about moving the core executor into `kubernetes_executor/__init__.py`, and moving the stuff we are pulling out into files under that new folder? I'm not sure I love the `kubernetes_executor_x.py` pattern we are establishing here.

Reply:
That to me felt a bit invasive, especially since executors are public-ish (people modify them with plugins or base new executors off of them).

We plan to move the executors to their own provider packages very soon so module names/locations will all be shuffled then. Can we agree to defer this piece of feedback until that time? I've already gotten feedback that this PR is too large, so if I could keep the two decoupled that'd be awesome!

Reply:
I'm not sure that new approach would have any more impact than the current one, on the public-ish-ness front. If folks are literally patching the executor, sure, but 🤷‍♂️.

My concern is how soon "very soon" is in reality. I'd hate to land 2.7 with this pattern in it. I'd be happier to say we do the refactoring in a follow up than tie it to the provider move.

Reply:
> I'm not sure that new approach would have any more impact than the current one, on the public-ish-ness front.

That's my bad, I should have been more clear. Let me take another hack at it: what I meant by that was let's not break things twice for users by changing the import path of the executor now and then changing it again when it's moved to it's own provider package. Not that it will be any different, just package the changes to paths all at once.

> I'd be happier to say we do the refactoring in a follow up than tie it to the provider move.
If you still disagree with the above then that's fair (though I'm curious to hear!) and I'd agree to cutting a ticket to move it after this PR :+1:

Reply:
> what I meant by that was let's not break things twice for users by changing the import path of the executor now

That's the beauty of it though, users would still:

```
from airflow.executors.kubernetes_executor import KubernetesExecutor
```

So the stuff that stays in the "main" file just moves on disk, how folks would import it wouldn't change at all. (We can ignore the stuff being moved like KubernetesJobWatcher, those are changing either way)

Am I overlooking something here (very possible)?

Reply:
> Am I overlooking something here (very possible)?

Ah nope, I see what you mean now! Yeah, I _think_ that would cover most interactions that users have with patching and base classing the executor. I'm happy to make those changes, but I'd request that it'd be in a follow-up PR.

Reply:
> I'm not sure I love the kubernetes_executor_x.py pattern

But I will note that I'd still name the module something quite similar to this, even in it's own dir. Since generic module names like `utils.py` can be difficult to work with due to the shared namespace.

Reply:
> That to me felt a bit invasive, especially since executors are public-ish (people modify them with plugins or base new executors off of them).

I think executor *user-facing behavior* is public but executor *code* is not public.

I think this was the consensus we arrived at, more or less, when @potiuk brought it up on the list, though I don't recall if it ever made it into "policy" via a PR.

So if you subclass an exectuor, you do so very much at your own risk.  And therefore we can change path of executor or helpers without worrying about backcompat, and refactor / move executor code at will as long as it doesn't break user-facing behavior (e.g. some radical breaking change w.r.t. user-supplied executor_config).